**Analysis for Data Set from Hacker News**

Here will be an analysis for some data from the website Hacker News. We will be observing if the tags "Ask HN" and "Show HN" receive more comments on average. We will also be looking to see if posts created at a certain time receive more comments on average.

In [2]:
from csv import reader

open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

# removing the header with the data column names
header = hn[0]
hn = hn[1:]

ask_posts = []
show_posts = []
other_posts = []

# create list within list of data based on whether
# the article title started with 'ask hn' or 'show hn'
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    if title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

# for 'ask hn' posts calculating the total amount of comments
# and average amount of comments
total_ask_comments = 0
avg_ask = 0
for row in ask_posts:
    val = int(row[4])
    avg_ask += 1
    total_ask_comments += val
    avg_ask_comments = total_ask_comments / avg_ask
    
print(total_ask_comments)
print(avg_ask_comments)

# for 'show hn' posts calculating the total amount of comments
# and average amount of comments
total_show_comments = 0
avg_show = 0
for row in show_posts:
    val = int(row[4])
    avg_show += 1
    total_show_comments += val
    avg_show_comments = total_show_comments / avg_show
    
print(total_show_comments)
print(avg_show_comments)

24483
14.038417431192661
11988
10.31669535283993


Based on my findings ask posts receive more comments than show posts with an average of 14.0 comments per item versus 10.3 comments per item. This could be because ask posts are explicitly seeking comments.

In [10]:
import datetime as dt

# since ask posts had a higher average comments only looking into them
# create a list within list with the first value of the list being
# the date with time and the second value being the amount of comments
result_list = []
for row in ask_posts:
    result_list.append([row[6],int(row[4])])

counts_by_hour = {}
comments_by_hour = {}

# formating the first value of each list and removing the hour element
# then counting the amount of posts by the number of posts
# and by the amount of comments
for row in result_list:
    full_date = dt.datetime.strptime(row[0], '%m/%d/%Y %H:%M')
    time_dt = dt.datetime.strftime(full_date, '%H')
    
    if time_dt in counts_by_hour:
        counts_by_hour[time_dt] += 1
        comments_by_hour[time_dt] += row[1]
    else:
        counts_by_hour[time_dt] = 1
        comments_by_hour[time_dt] = row[1]

# calculating the average per hour and average per comments
avg_by_hour = []
for avg_hour in counts_by_hour:
    avg_val = comments_by_hour[avg_hour] / counts_by_hour[avg_hour]
    avg_by_hour.append([avg_hour, avg_val])

# created a new list where the order of the lists were swapped
swap_avg_by_hour = []
for row in avg_by_hour:
    row_1 = row[1]
    row_2 = row[0]
    swap_avg_by_hour.append([row_1, row_2])

# sorted the list by the highest value to the smallest
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print("Top 5 Hours to Ask Posts Comments")
for row in sorted_swap[:5]:
    template = "{}: {:.2f} average comments per post"
    # format the average hour to be an hour
    get_hrs = dt.datetime.strptime(row[1], '%H')
    hrs = dt.datetime.strftime(get_hrs, '%H:%M')
    print(template.format(hrs, row[0]))

Top 5 Hours to Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


Based on my findings the best time to post an ask post is at 3pm by far, but 